ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/MetaStoreCompactorThread.java

```diff
@@ -57,15 +50,15 @@ public class MetaStoreCompactorThread extends CompactorThread implements MetaSto
 
   protected TxnStore txnHandler;
   protected ScheduledExecutorService cycleUpdaterExecutorService;
-
-  private Optional<Cache<String, TBase>> metaCache = Optional.empty();
+  protected CacheContainer cacheContainer;
 
   @Override
   public void init(AtomicBoolean stop) throws Exception {
     super.init(stop);
 
     // Get our own instance of the transaction handler
     txnHandler = TxnUtils.getTxnStore(conf);
+    cacheContainer = new CacheContainer();
```

Comment:
shouldn't it be a singleton? why do we need multiple cache instances for Initiator & Cleaner?

Reply:
As discussed, the cached instances are invalidated both on Initiator and Cleaner side. There might also be staleness of table object if properties of the table are changed in the time between Initiator & Cleaner.